backend/src/controller/chatController.ts

```diff
@@ -21,27 +23,82 @@ import { Defence } from '@src/models/defence';
 import { EmailInfo } from '@src/models/email';
 import { LEVEL_NAMES } from '@src/models/level';
 import { chatGptSendMessage } from '@src/openai';
+import { pushMessageToHistory } from '@src/utils/chat';
 
 import { handleChatError } from './handleError';
 
+function combineChatDefenceReports(
+	reports: ChatDefenceReport[]
+): ChatDefenceReport {
+	const combinedReport: ChatDefenceReport = {
+		blockedReason: reports
+			.filter((report) => report.blockedReason !== null)
+			.map((report) => report.blockedReason)
+			.join('\n'),
+		isBlocked: reports.some((report) => report.isBlocked),
+		alertedDefences: reports.map((report) => report.alertedDefences).flat(),
+		triggeredDefences: reports.map((report) => report.triggeredDefences).flat(),
+	};
+	return combinedReport;
+}
+
+function getChatHistoryUserMessages(
+	message: string,
+	transformedMessage: string | null
+): ChatHistoryMessage[] {
+	if (transformedMessage) {
+		// if message has been transformed
+		return [
+			// original message
+			{
+				completion: null,
+				chatMessageType: CHAT_MESSAGE_TYPE.USER,
+				infoMessage: message,
+			},
+			// transformed message
+			{
+				completion: {
+					role: 'user',
+					content: transformedMessage,
+				},
+				chatMessageType: CHAT_MESSAGE_TYPE.USER_TRANSFORMED,
+			},
+		];
+	} else {
+		// not transformed, so just return the original message
+		return [
+			{
+				completion: {
+					role: 'user',
+					content: message,
+				},
+				chatMessageType: CHAT_MESSAGE_TYPE.USER,
+			},
+		];
+	}
+}
+
 // handle the chat logic for level 1 and 2 with no defences applied
 async function handleLowLevelChat(
 	message: string,
 	chatResponse: ChatHttpResponse,
 	currentLevel: LEVEL_NAMES,
 	chatModel: ChatModel,
 	chatHistory: ChatHistoryMessage[],
-	defences: Defence[],
-	sentEmails: EmailInfo[]
+	defences: Defence[]
 ): Promise<LevelHandlerResponse> {
+	let updatedChatHistory = [...chatHistory];
+	const chatHistoryUserMessages = getChatHistoryUserMessages(message, null);
+	chatHistoryUserMessages.forEach((message) => {
+		updatedChatHistory = pushMessageToHistory(updatedChatHistory, message);
+	});
```

Comment:
Avoid `let` declarations where possible, to aid in immutability. Try this:

```
const updatedChatHistory = getChatHistoryUserMessages(message, null)
  .reduce(pushMessageToHistory, chatHistory);
```

Reply:
Lovely! I still need to get into the habit of using `reduce`